python/tk_multi_launchapp/software_entity_launcher.py

```diff
@@ -246,55 +232,170 @@ def _build_register_command_data(self, display_name, icon, engine, path, args, v
-                    self._tk_app.sgtk, self._tk_app.context, engine
+            # No application path(s), no launch command(s) ....
+            self._tk_app.log_warning(
+                "No application path or Toolkit engine specified for Software %s. "
+                "Cannot create launch commands associated with this entity." % display_name
+            )
+
+        # Check if there are icons to download and whether we're in an appropriate
+        # environment to do so.
+        if download_icon_for_commands and self._tk_app.engine.has_ui:
+            if sg_software_type and sg_software_id:
+                # Import sgutils after ui has been confirmed because it has dependencies on Qt.
+                shotgun_data = sgtk.platform.import_framework(
+                    "tk-framework-shotgunutils", "shotgun_data"
                 )
-                if launcher:
-                    # Get a list of SoftwareVersions for this engine and use that
-                    # data to construct a launch command to register.
-                    sw_versions = launcher.scan_software(versions, display_name, icon)
-                    for swv in sw_versions:
-                        commands.append({
-                            "display_name": swv.display_name, "icon": swv.icon,
-                            "engine": engine, "path": swv.path, "args": args,
-                            "version": swv.version
-                        })
-                else:
-                    self._tk_app.log_error(
-                        "Engine %s does not support scanning for software versions." %
-                        engine
-                    )
-            except Exception, e:
+
+                # Download the Software thumbnail source from Shotgun and cache for reuse.
+                self._tk_app.log_debug("Downloading app icon for %s from %s %s ..." %
+                    (display_name, sg_software_type, sg_software_id)
+                )
+                local_icon = shotgun_data.ShotgunDataRetriever.download_thumbnail_source(
+                    sg_software_type, sg_software_id, self._tk_app
+                )
+                self._tk_app.log_debug("... download complete: %s" % local_icon)
+
+                # Update the launch commands with the local icon value.
+                [cmd.update({"icon": local_icon}) for cmd in download_icon_for_commands]
+
+            else:
                 self._tk_app.log_warning(
-                    "Cannot determine executable paths for %s: %s" %
-                    (engine, e)
+                    "Missing entity information to download source thumbnails. Expecting "
+                    "valid values for entity type (got %s) and id (got %s). Related icons "
+                    "may not display correctly." % (sg_software_type, sg_software_id)
                 )
 
         return commands
+
+    def _engine_launcher_software_versions(self, engine, default_name, default_icon, versions=None):
```

Comment:
i am missing a verb - what does this do? maybe slap on a get to make it `_get_engine_launcher_software_versions`? Or simply `_scan_for_software`?